lib/vsdata.cpp

```diff
@@ -47,6 +47,60 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int order = gf.FESpace()->GetMaxElementOrder();
+
+   // check for integral elements
+   const int dim = mesh->Dimension();
+   const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+   if (fec && fec->GetMapType(dim) == FiniteElement::INTEGRAL)
+   {
+      cout << "Warning: integral elements are non-polynomial in the physical space,\n"
+           << "         consider increasing the refinement by the key 'o'."
+           << endl;
+   }
+
+   return std::max(order, 1);
+}
+
+int VisualizationSceneScalarData::GetAutoRefineFactor()
+{
+   const int dim = mesh->Dimension();
+   const int ne = (dim == 3)?(mesh->GetNBE()):(mesh->GetNE());
+
+   // determine the refinement based on the order of the mesh and grid function
+   int order_ref = GetFunctionAutoRefineFactor();
+
+   // mesh
+   const FiniteElementSpace *nfes = mesh->GetNodalFESpace();
+   if (nfes)
+   {
+      const int order = nfes->GetMaxElementOrder();
+      order_ref = std::max(order_ref, order);
+   }
+
+   // limit the total number of elements
+   int auto_ref_surf_elem = ne*(order_ref+1)*(order_ref+1);
+   auto_ref_surf_elem = std::min(std::max(auto_ref_surf_elem,
+                                          auto_ref_min_surf_elem), auto_ref_max_surf_elem);
```

Comment:
Having a `min_surf_elem` means that for a trivially small dataset (1 element) this will still try to refine to 100000 elements? (I ran this on `streams/quad.saved` and it defaulted to 17 subdivision factors 👀). Maybe it isn't a huge deal but it seems unnecessary?

Why not just use `min(auto_ref_surf_elem, auto-ref_max_surf_elem)`? Is it because we think that the estimate is too conservative?

Reply:
The algorithm always used some minimum, which was 20k formerly, but we decided to increase it. Go to the beginning of the PR ☝️ .

Reply:
100k sounds like a lot, but it is nothing on modern GPUs 😉

Reply:
Btw, `streams/quad.saved` always run with 16x refinement if you look back :wink: . We are just showing off when we can 😎

Reply:
All these `*surf*` names should be replaced with something that uses `vert` instead -- for a given refinement value `ref`, the formula `ne*(ref+1)*(ref+1)` gives the number of vertices that will be used for `ref` subdivisions.

Reply:
On quad or hex elements, "linear" elements are actually bi-linear and tri-linear, respectively. I other words a solution on a quad can have nonlinear terms like $x y$ -- this is the main reason to allow refinements even for linear elements. The final `ref` is also limited to be $\le 16$, so for 1 element you will get at most $16 \times 16 = 256$ quads using $17 \times 17 = 289$ vertices.

Reply:
> All these `*surf*` names should be replaced with something that uses `vert` instead -- for a given refinement value `ref`, the formula `ne*(ref+1)*(ref+1)` gives the number of vertices that will be used for `ref` subdivisions.

It is matter of interpretation 🤔 . The algorithm can be interpreted either as:

- it iterates in number of elements with the loop being advanced by one iteration, stepping to the refinement such that the number of elements is the closest lower or equal value to `auto_ref_surf_elem`,
- or that it iterates in number vertices, stepping to the refinement such that the number of vertices is the closest higher value than `auto_ref_surf_elem`.

These two interpretations are equivalent, but the current naming suggests it was meant as the former one, so I would keep it that way.

Reply:
100k is not a lot but if you are trying to process many different images of small solutions (e.g. a large parameter sweep) it could add up. But maybe that is not a common enough use-case.

Just to make sure we're all on the same page - the intended `auto_ref_surf_elem` of this algorithm (number of elements we are trying to get "close" to) is $ne*order^2$ where $order=max(order_{mesh}, order_{gf})$. The `+1` is only there because of incrementation in the `ref++` loop, as @najlkin explained in the previous thread: https://github.com/GLVis/glvis/pull/285#discussion_r1695491018). So for linear elements this is not intended to refine (appropriate for linear tris/tets but not quads/hexes as Veselin is pointing out).

Previously I thought this algorithm was trying to reach $ne*(order+1)^2$, which would subdivide e.g. a linear solution on a triangular mesh 4x (even though it shouldn't be subdivided at all).

I don't think there is a perfect solution to this, but based on my understanding now and Veselin's comment, I could see a case for increasing `auto_ref_surf_elem` if the mesh is a quad or hex.

Reply:
I don't want to drag this out though, and I think this algorithm is definitely an improvement from the previous one, so I am okay with merging whenever!

Reply:
What should be the refinement should not be calculated here, but in `GetFunctionAutoRefineFactor()`. After the discussion on Slack, we decided to reduce it to `order_ref=order`, where @v-dobrev was actually pushing that. Please do not open it again 😟 . I think he was just saying that increasing the refinement when it is cheap (under 100k limit) makes sense, because the `order_ref` does not guarantee perfect visualization, especially with bilinear quads and so on.

Reply:
Okay, I'm all on-board!